### PR TITLE
Alistair/i2c mux

### PR DESCRIPTION
**Pull request by alistair23**
### Pull Request Overview

To allow an I2C bus to support devices created by the board **and** still be exposed to userspace let's convert the I2CMasterDriver to use an `I2CDevice` instead of the physical I2C hardware.

This allows us to use the Mux with userspace applications and kernel uses.

### Testing Strategy

Communicating with a I2C device from userspace while still supporting kernel devices.

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

### CODE REVIEW

**Review by lschuermann — APPROVED** [review]
This makes sense to me, in particular also the part about `I2CDevice`s being able to change their address at runtime (e.g., to support peripherals where the address is flashed into the device on initial configuration). I'm perhaps slightly concerned that this will make it easier to create address conflicts on a given bus (as the kernel can't know what userspace is doing on this bus), but presumably the kernel has some amount of trust towards the app when it gives it access to a shared I2C bus, and later changes could introduce some optional filtering infrastructure.

**Review by bradjc — CHANGES_REQUESTED** [review]
I feel like this is digging us a deeper hole. `I2CDevice` is explicitly supposed to restrict a user to a single I2C device (or really a single I2C address).

I2C is an example of why virtualizing with a different type is problematic.

I think we need to add a different virtualization layer.

File: boards/components/src/i2c.rs [hunk 1]
```diff
@@ -173,3 +191,56 @@ impl<I: 'static + i2c::I2CMasterSlave<'static>> Component for I2CMasterSlaveDriv
         i2c_master_slave_driver
     }
 }
+
+pub struct I2CMasterDriverComponent<I: 'static + i2c::I2CMaster<'static>> {
```
Comment (bradjc) [inline boards/components/src/i2c.rs:195]:
```suggestion
pub type I2CMasterDriverComponentType<I> = capsules_core::i2c_master::I2CMasterDriver<I2CDevice<I>>;

pub struct I2CMasterDriverComponent<I: 'static + i2c::I2CMaster<'static>> {
```

**Comment by alistair23** [comment]
> I feel like this is digging us a deeper hole. `I2CDevice` is explicitly supposed to restrict a user to a single I2C device (or really a single I2C address).

This doesn't change that.

If a board only creates an `I2CDevice` then that is all that is exposed to userspace. Userspace still can't change the address.

If a board designer wants to expose an `I2CDevice` **and** the I2C bus, this allows them to expose both. It also still allows only exposing the bus and no `I2CDevice`. It just allows the board configuration to be more configurable.

I don't see how a different virtuliser would help here. What do you think it should virtulise?

File: kernel/src/hil/i2c.rs [hunk 2]
```diff
@@ -230,6 +230,7 @@ pub trait I2CDevice {
     fn write(&self, data: &'static mut [u8], len: usize) -> Result<(), (Error, &'static mut [u8])>;
     fn read(&self, buffer: &'static mut [u8], len: usize)
         -> Result<(), (Error, &'static mut [u8])>;
+    fn set_address(&self, addr: u8);
```
Comment (bradjc) [inline kernel/src/hil/i2c.rs:233]: We can't add this API. The address is fixed, and grants the holder of an `I2CDevice` access to only the specified device (address).

> /// It gives an interface for communicating with a specific I2C
> /// device.

Comment (alistair23) [reply kernel/src/hil/i2c.rs:233]: What is the device address changes though? It's not uncommon to have programmable I2C addresses

Comment (bradjc) [reply kernel/src/hil/i2c.rs:233]: I think either a capsule could use the `I2CMaster` interface, or a shim layer that tracks the address changes below the `I2CDevice` would work.

**Comment by alistair23** [comment]
I don't think there is a better way to do this.

The other approach is to modify the `virtual_i2c`. We could have `MuxI2C` implement `i2c::I2CMaster` so then the `i2c_master::I2CMasterDriver` can just use `MuxI2C` directly.

That would involve adding something like `i2c_bus: OptionalCell<&'a I2CMaster<'a, I>>` to `MuxI2C` and tracking operations and addresses from the `i2c::I2CMaster` implementation. Then `do_next_op()` would call the I2C hardware based on pending operations.

In the long run though, that seems even worse. Then there is no way to filter or prevent userspace from accessing other I2C devices. This current implementation will allow us to restrict access to I2C devices from userspace as required. We can also check for conflicts in `set_address()`, Something which I think will be useful.

**Comment by bradjc** [comment]
Can you outline what you are trying to achieve?



It seems like something like this might work? Feels like we need multiple virtualization layers to support both kernel and userspace and multiple userspace apps.

```
  I2CDevice
  --------
   I2CMux                 Userspace
-----------              -----------
Virtual_I2C              Virtual_I2C
------------------------------------
               I2CMux
------------------------------------
             I2CMaster
```

**Comment by alistair23** [comment]
I'm trying to use a single I2C bus for some `I2CDevice`s and also expose the bus to userspace.

Another virtuliser layer to virtulise `i2c::I2CMaster` directly has too much overhead. We end up with 2 layers of virutlisation, the second only virtulising two users (`MuxI2C` and `I2CMasterDriver`)

**Comment by bradjc** [comment]
> has too much overhead

In terms of latency? Or memory use?

---


It sounds like you need a new capsule, once which uses the `I2CMaster` interface, but then allows the kernel to create `I2CDevice`s for its own use and exposes the userspace interface you want.

**Comment by alistair23** [comment]
> In terms of latency? Or memory use?

Memory usage

> It sounds like you need a new capsule, once which uses the `I2CMaster` interface, but then allows the kernel to create `I2CDevice`s for its own use and exposes the userspace interface you want.

I don't think that's a great idea though, see https://github.com/tock/tock/pull/3867#issuecomment-1955390203

File: kernel/src/hil/i2c.rs (hunk 2 above)
Comment (alevy) [reply kernel/src/hil/i2c.rs:233]: I agree with @bradjc, for this abstraction, anyway, the whole point of the `I2CDevice` trait is that it isolates drivers talking to different peripherals over the same I2C bus. If a client of `I2CDevice` can modify the address arbitrarily, it completely negates the point of this abstraction.

There _could_ be a _different_ interface. For example, an extension trait might work, but I'd really want to understand the use case. Is the idea that some application/driver would change the I2C device dynamically after boot? What's a specific case for this?

Comment (bradjc) [reply kernel/src/hil/i2c.rs:233]: The goal is to be able to share an I2C bus with multiple kernel drivers and with userspace. Something like:

```
                             App1    App2   
                            =====syscall=====
Sensor         Screen        I2CMasterDriver
---------     ---------     -----------------
I2CDevice     I2CDevice         I2CDevice(?)
---------------------------------------------
                 Virtual I2C
---------------------------------------------
                   I2C bus
```

**Comment by lschuermann** [comment]
How about a second instance of the system call driver? AFAIK they were always assumed to be duplicable and can be instantiated at a second syscall driver number. This would solve many such problems across subsystems, is something that downstream users are likely to do anyways, and may motivate userspace APIs that can be supplied an alternative driver number.

**Comment by bradjc** [comment]
> How about a second instance of the system call driver? AFAIK they were always assumed to be duplicable and can be instantiated at a second syscall driver number. This would solve many such problems across subsystems, is something that downstream users are likely to do anyways, and may motivate userspace APIs that can be supplied an alternative driver number.

Yes that would be no problem, but userspace would only have access to a single I2C device (address) without a different virtualizer.

**Comment by alistair23** [comment]
> What should we do with this PR? Is it helpful to serve as a placeholder? Or close it?

I think this is something that needs to be supported. Otherwise we loose the entire I2C bus when one `I2CDevice` is created.

Maybe some else can have a go doing it in a way you approve of?

**Comment by alistair23** [comment]
> A capsule that exposes `I2CDevice` to userspace would work.

How does that fix the issue?

**Comment by bradjc** [comment]
The existing virtual_i2c layer could be used to provide `I2CDevice`s for both kernel drivers and userspace.

**Comment by alistair23** [comment]
I don't see how that helps the original problem though. How is the I2C bus then exposed to userspace?

**Comment by bradjc** [comment]
I don't think it makes sense to provide an I2CDevice to a kernel driver, and then also provide access to the same device (address) via a low level bus interface to userspace. I mean it's possible using another virtualizer, but I don't think it fits the Tock ethos of isolation to allow userspace to interfere with the kernel.

**Comment by alistair23** [comment]
> I don't think it makes sense to provide an I2CDevice to a kernel driver, and then also provide access to the same device (address) via a low level bus interface to userspace. I mean it's possible using another virtualizer, but I don't think it fits the Tock ethos of isolation to allow userspace to interfere with the kernel.

This is what the very first comment was also talking about: https://github.com/tock/tock/pull/3867#pullrequestreview-1888441318

This PR isn't for providing an `I2CDevice` to the kernel and userspace. That already works in Tock today. The idea is to support having multiple devices on an I2C bus. Some represented as `I2CDevice`s and some as just accessible via `I2CMasterDriver` from userspace. That way we don't need to support every single type of I2C device in the kernel.

If this was merged I would not have bothered with https://github.com/tock/tock/pull/3881 for example

As [Leon](https://github.com/tock/tock/pull/3867#pullrequestreview-1888441318) and [I](https://github.com/tock/tock/pull/3867#issuecomment-1955390203) have pointed out, it would be pretty easy to add some filtering on top of this implementation, to avoid conflicts between userspace and the kernel.

I don't see how a "capsule that exposes I2CDevice to userspace" helps address the limitation

**Comment by bradjc** [comment]
My proposed workflow:
1. main() instantiates `I2CDevice`s for each I2C device on the board.
2. main provides some of the `I2CDevice`s to in-kernel drivers.
3. main instantiates one or more of my proposed `i2c_device_driver.rs` userspace drivers.
4. main provides each with a `I2CDevice`.
5. main exposes the i2c_device_drivers to userspace with different syscall driver numbers.

This supports I2C devices from both kernel and userspace. As mentioned, I don't see a case where we want to allow userspace to access I2C addresses that the kernel is using. The filtering we have is the I2CDevice abstraction.

**Comment by alistair23** [comment]
> My proposed workflow:
> 
>     1. main() instantiates `I2CDevice`s for each I2C device on the board.
> 
>     2. main provides some of the `I2CDevice`s to in-kernel drivers.
> 
>     3. main instantiates one or more of my proposed `i2c_device_driver.rs` userspace drivers.
> 
>     4. main provides each with a `I2CDevice`.
> 
>     5. main exposes the i2c_device_drivers to userspace with different syscall driver numbers.
> 
> 
> This supports I2C devices from both kernel and userspace. As mentioned, I don't see a case where we want to allow userspace to access I2C addresses that the kernel is using. The filtering we have is the I2CDevice abstraction.

How can userspace know which devices have been exposed? Is userspace supposed to just brute force syscall numbers? How does userspace know which syscall corresponds to which device? How do we keep that consistent across boards?

Also this greatly increases the complexity of boards. If a device wants to communicate with an off-board I2C device we need to create a custom board and then re-compile and re-flash the kernel. That would be a large scale up of https://github.com/tock/tock/pull/3873 to support a range of I2C device combinations.

I feel that your proposed approach reduces a lot of the OS abstraction flexibility. Unless I'm misunderstanding the idea

**Comment by alevy** [comment]
I'm quite confused about what this PR is trying to accomplish. I'm particularly failing to square some of the comments in the thread with the fact that this PR adds a `set_address` to the `I2CDevice` trait. But also, the PR description kind of starts from the middle.

What is the _specific_ use case? What is the happy end-goal?

_Specifics_:

1. What does a user-space process get to do with an `I2C` system call driver. Currently, it gets to send/receive I2C to arbitrary I2C devices on one bus. There is an invariant that no other kernel component is using any device on the same bus (because multiplexing only happens through virtual `I2CDevice`s). Multiple processes can interfere with each other (this is _already_ a problem, and realistically, if this is ever used in practice, it should probably be restricted to a single process).
2. What's an example scenario that needs this (what specific problem does this solve)? If the problem is "I want to be able to talk to a raw I2C bus from userspace, e.g. for development or debugging" this is already possible. If the problem is "I want to be able to talk to the same I2C device concurrently from both the kernel and userspace" I'm skeptical there is a path for supporting that use case upstream (it really goes against much of kind of isolation upstream is _trying_ to achieve). If the problem is "I want to be able to talk to devices from processes which are on the same bus as other devices which are used in kernel components, but are distinct devices" this PR doesn't actually enforce or accomplish that. @bradjc's suggestion of an driver that "owned" a bunch of (virtual) `I2CDevice`s and allowed processes to talk specifically to those would work (though I'm empathetic, at least in principle, to the discoverability concerns). Maybe there is another mechanism.

If there is some gap between what is reasonable to accomplish in a single PR some happy end that is being implied here, what is that? What are the steps to get there?

**Comment by alistair23** [comment]
> What is the _specific_ use case? What is the happy end-goal?

The end goal is what the PR description says

To allow an I2C bus to support devices created by the board and still be exposed to userspace let's convert the I2CMasterDriver to use an I2CDevice instead of the physical I2C hardware.

This allows us to use the Mux with userspace applications and kernel uses.

If you have an I2C bus on a board, with multiple I2C devices connected to a physical bus. Some of those devices are supported by kernel capsules (say a https://github.com/tock/tock/blob/master/capsules/extra/src/bmp280.rs) and some aren't (such as a GNSS device). Currently you have to set up Tock such that every single physical I2C device has a capsule written for it. Otherwise you can't access it from user space (unless you use no kernel capsules).

This PR changes the virtulisation such that a I2C bus can have both kernel supported capsules **AND** still expose the other devices to userspace.


> * What does a user-space process get to do with an `I2C` system call driver. Currently, it gets to send/receive I2C to arbitrary I2C devices on one bus. There is an invariant that no other kernel component is using any device on the same bus (because multiplexing only happens through virtual `I2CDevice`s). Multiple processes can interfere with each other (this is _already_ a problem, and realistically, if this is ever used in practice, it should probably be restricted to a single process).
> 

Yes. This PR does not address the existing issues with the I2C syscall drivers. The drivers are already used and enabled by default on a range of Tock boards, so I'm not sure why this PR is the issue. If there are multiple apps running on any Tock board using an I2C driver they can conflict with each other. They can also conflict with the kernel as well.

The [nRF](https://github.com/tock/tock/blob/master/boards/nordic/nrf52840dk/src/main.rs#L790) as well as pretty much every board already has this problem. This PR doesn't change that in anyway.

One thing that this PR does do, is move in a direction where we can in the future add filtering to the I2C accesses. That would then allow us to ensure no two users are communicating with the same address. This PR does **NOT** do that, it just makes that easier in the future.

Just to be clear, this PR doesn't fix or worsen the existing power of the I2C driver to access the entire bus. It just allows a board design to **also** use `I2CDevice`s on the bus.

> 2\. What's an example scenario that needs this (what specific problem does this solve)? If the problem is "I want to be able to talk to a raw I2C bus from userspace, e.g. for development or debugging" this is already possible. If the problem is "I want to be able to talk to the same I2C device concurrently from both the kernel and userspace" I'm skeptical there is a path for supporting that use case upstream (it really goes against much of kind of isolation upstream is _trying_ to achieve). If the problem is "I want to be able to talk to devices from processes which are on the same bus as other devices which are used in kernel components, but are distinct devices" this PR doesn't actually enforce or accomplish that. @bradjc's suggestion of an driver that "owned" a bunch of (virtual) `I2CDevice`s and allowed processes to talk specifically to those would work (though I'm empathetic, at least in principle, to the discoverability concerns). Maybe there is another mechanism.

"I want to be able to talk to devices from processes which are on the same bus as other devices which are used in kernel components, but are distinct devices" is exactly what this PR allows. Before https://github.com/tock/tock/pull/3881 I was using this PR to get NMEA sentences from a I2C device while also allowing the kernel to communicate with other I2C devices.

Why do you think this PR doesn't accomplish that?

Basically the idea is that Tock can be flashed to a device. I can plug in an I2C device and I don't have to re-flash the kernel. Instead I can install an app that can communicate with that device via I2C.

Considering how hard it is to support new I2C devices (see https://github.com/tock/tock/pull/3881) I think this is a reasonable approach.

I think @bradjc's suggestion has a few issues, as I pointed out in https://github.com/tock/tock/pull/3867#issuecomment-1994194696. I might be misunderstanding the suggestion, but I haven't heard a response explaining what I'm missing.

**Comment by bradjc** [comment]
> Yes. This PR does not address the existing issues with the I2C syscall drivers. The drivers are already used and enabled by default on a range of Tock boards, so I'm not sure why this PR is the issue.


Currently we support two different I2C setups, if you will:

1. I2C devices are supported in the kernel, and userspace only has access through dedicated drivers. Semantically, the kernel owns the bus. This matches the Tock ethos of isolating the resource with the kernel.
2. The I2C bus is entirely exposed to userspace. Semantically, userspace owns the I2C bus. This provides no guarantees on isolation. In practice, this is for debugging and development boards. I can't see how a production kernel would ever want to expose this level of control userspace.

This PR merges the two, and we need to compare to setup #1 as the baseline (not #2). To maintain isolation guarantees for the kernel, we need to do this very carefully. That the "debugging" I2C setup (setup #2) allows for resource conflict does not mean we can compromise setup #1. We need to ensure that the kernel can maintain strong I2C isolation guarantees as that is what existing I2C-in-kernel users expect.

**Comment by bradjc** [comment]
> How can userspace know which devices have been exposed? Is userspace supposed to just brute force syscall numbers? How does userspace know which syscall corresponds to which device? How do we keep that consistent across boards?

I don't know how userspace would know the syscall driver numbers, but I assume it would work similarly to how userspace would know what I2C addresses are available and what those addresses correspond to. This is a fundamental issue with exposing any low-level/generic interface.

> Also this greatly increases the complexity of boards. If a device wants to communicate with an off-board I2C device we need to create a custom board and then re-compile and re-flash the kernel. That would be a large scale up of #3873 to support a range of I2C device combinations.

I don't think we can have everything. The ideal solution might support these features:

1. Userspace can access an I2C bus directly, for accessing arbitrary devices added to the bus.
2. The kernel can use the I2C bus for well-known devices on the board.
3. Tock provides robust isolation and guarantees userspace cannot interfere with an I2C device the kernel owns.
4. Minimal memory overhead for virtualization.

If feature 2 is enabled, then feature 3 is also a hard requirement for upstream code. Out-of-tree code of course does not need to uphold 3.

- Set (1, 2, 3) is possible by adding a second virtualization layer.
- Set (2, 3, 4) is possible by instantiating `I2CDevice`s for each I2C device.
- Set (1, 2) is possible with this PR, but it doesn't enable feature 3.

**Comment by alistair23** [comment]
> 1. I2C devices are supported in the kernel, and userspace only has access through dedicated drivers. Semantically, the kernel owns the bus. This matches the Tock ethos of isolating the resource with the kernel.

Agreed. This PR doesn't change this setup

> 2\. The I2C bus is entirely exposed to userspace. Semantically, userspace owns the I2C bus. This provides no guarantees on isolation. In practice, this is for debugging and development boards. I can't see how a production kernel would ever want to expose this level of control userspace.

Yep. This is what a lot of Tock boards do today.

I see it as this PR changes this option to be a little more flexible. It allows exposing the I2C bus to user space (as most boards do), while also having dedicated drivers on the same bus.

A production kernel probably does not want to do this. Although this PR does move towards address filtering which might change that.

If the issue is specifically the `set_address()` it could be taken out of the HIL. Or there could be a bool set at `I2CDevice` creation to allow dynamic addresses.

> I don't think we can have everything. The ideal solution might support these features:
> 
>     1. Userspace can access an I2C bus directly, for accessing arbitrary devices added to the bus.
> 
>     2. The kernel can use the I2C bus for well-known devices on the board.
> 
>     3. Tock provides robust isolation and guarantees userspace cannot interfere with an I2C device the kernel owns.
> 
>     4. Minimal memory overhead for virtualization.

Wouldn't a simple filter on top of this get 1, 2, 3 and 4? We could just check if a device is already in use before allowing `set_address()` to succeed

**Comment by bradjc** [comment]
> If the issue is specifically the `set_address()` it could be taken out of the HIL.

This is the issue. We absolutely cannot add this API to `I2CDevice` HIL.

> Wouldn't a simple filter on top of this get 1, 2, 3 and 4? We could just check if a device is already in use before allowing `set_address()` to succeed

Maybe. I'm concerned it would be fragile and easy to get wrong. Unless there is a way to use the type system to ensure that all `I2CDevice`s are accurately represented in the filter.

**Comment by alistair23** [comment]
> This is the issue. We absolutely cannot add this API to `I2CDevice` HIL.

Would a public function not in the HIL work?

> Maybe. I'm concerned it would be fragile and easy to get wrong. Unless there is a way to use the type system to ensure that all `I2CDevice`s are accurately represented in the filter.

If there is a general consensus that this can be accepted with filtering I could add it. I would rather no add it if this isn't going to be accepted though.

It should be as simple as (untested)

```rust
fn set_address_check(&self, new_addr: u8) -> bool {
   self
        .i2c_devices
        .iter()
        .find(|node| node.addr.get() == new_addr).is_none()
}

fn set_address(&self, addr: u8) -> Result<> {
    if self.mux.set_address_check(addr) {
        self.addr.set(addr);
        Ok(())
    } else {
        Err(())
    }
}
```

That should avoid conflicts with `I2CDevices`

**Comment by bradjc** [comment]
I think a design where the I2C mux exposes the "rest" of the I2C bus as a separate object would be workable. I don't think it should be called `I2CDevice` as that implies a single device. But the virtualizer could support a separate interface for all I2C addresses that are not used by `I2CDevice`s.

I think that would balance the competing concerns.

**Comment by alistair23** [comment]
> I think a design where the I2C mux exposes the "rest" of the I2C bus as a separate object would be workable. I don't think it should be called `I2CDevice` as that implies a single device. But the virtualizer could support a separate interface for all I2C addresses that are not used by `I2CDevice`s.
> 
> I think that would balance the competing concerns.

I pushed an update that I think addresses this. It builds but is untested, hence the draft status. Let me know if it is in the right direction and I'll test it

File: capsules/core/src/virtualizers/virtual_i2c.rs [hunk 3]
```diff
@@ -17,10 +17,12 @@ use kernel::utilities::cells::{OptionalCell, TakeCell};
 pub struct MuxI2C<'a, I: i2c::I2CMaster<'a>, S: i2c::SMBusMaster<'a> = NoSMBus> {
     i2c: &'a I,
     smbus: Option<&'a S>,
-    i2c_devices: List<'a, I2CDevice<'a, I, S>>,
+    i2c_devices: List<'a, I2CDevice<'a, I, false, S>>,
```
Comment (bradjc) [inline capsules/core/src/virtualizers/virtual_i2c.rs:20]: Rather than adding a bool, I think we should just have a separate object in the `MuxI2C` struct. The "rest of the bus" isn't really a device.

File: capsules/core/src/virtualizers/virtual_i2c.rs [hunk 4]
```diff
@@ -307,9 +398,17 @@ impl<'a, I: i2c::I2CMaster<'a>> i2c::I2CDevice for I2CDevice<'a, I> {
     }
 }
 
+impl<'a, I: i2c::I2CMaster<'a>> i2c::I2CMultiDevice for I2CDevice<'a, I, true, NoSMBus> {
```
Comment (bradjc) [inline capsules/core/src/virtualizers/virtual_i2c.rs:401]: Why not just use `I2CMaster`?

File: capsules/core/src/i2c_master.rs [hunk 5]
```diff
@@ -37,14 +37,14 @@ struct Transaction {
     read_len: OptionalCell<usize>,
 }
 
-pub struct I2CMasterDriver<'a, I: i2c::I2CMaster<'a>> {
+pub struct I2CMasterDriver<'a, I: i2c::I2CMultiDevice> {
```
Comment (bradjc) [inline capsules/core/src/i2c_master.rs:40]: Then we wouldn't need to change this.

**Review by bradjc — COMMENTED** [review]
I've been thinking more about this and I think this idea is very workable. I do think we want a clear distinction between an I2CDevice and the "remaining" I2C bus.

File: capsules/core/src/virtualizers/virtual_i2c.rs (hunk 3 above)
Comment (alistair23) [reply capsules/core/src/virtualizers/virtual_i2c.rs:20]: We do add a `i2c_bus` object to the `MuxI2C` object.

The bool is just to reuse the existing `I2CDevice` logic instead of copying everything

File: capsules/core/src/virtualizers/virtual_i2c.rs (hunk 4 above)
Comment (alistair23) [reply capsules/core/src/virtualizers/virtual_i2c.rs:401]: I feel the layering is a little weird using `I2CMaster`. As the `I2CDevice` layers on top of `I2CMaster` it feels a little cleaner to then also layer the "rest of the bus" on `I2CMaster` as well.

To me it also makes it clearer that we only have access to some of the bus, instead of full access that `I2CMaster` has.

By using `I2CMultiDevice` we can piggy back on the existing `I2CDevice` support. Plus then we need to check the address less often, only when setting it instead of on every operation

File: capsules/core/src/virtualizers/virtual_i2c.rs (hunk 3 above)
Comment (bradjc) [reply capsules/core/src/virtualizers/virtual_i2c.rs:20]: I think its ok to copy to get the new functionality and have it be clear and maintainable.

**Comment by alistair23** [comment]
> I've been thinking more about this and I think this idea is very workable. I do think we want a clear distinction between an I2CDevice and the "remaining" I2C bus.

Ready to go then

**Comment by alistair23** [comment]
Ping!

**Comment by alistair23** [comment]
Ping!

**Comment by alistair23** [comment]
Ping!

**Comment by alistair23** [comment]
Ping!

**Comment by alistair23** [comment]
Ping!

**Comment by alistair23** [comment]
Ping!

**Comment by lschuermann** [comment]
@bradjc You seemed to have the most involvement in the discussion so far, let me know if I should reassign myself or someone else.

**Comment by ppannuto** [comment]
I do think this is still something we want (even though we lost track of it, sorry :/).

Changing the internal `I2CMasterDriver` to be generic over an `I2CMultiDevice` instead of over an `I2CMaster` feels a bit odd. The artifact of that decision which feels particularly weird are [the changes to the internal operation from just `read` to `enable; set_addr; read`](https://github.com/tock/tock/pull/3867/files#diff-2f7355a91724ede0e050c1e88f926d74556686ad391d1540e3083e2f2077c1a1R92-R106).

Conceptually, I think that lower-level I2CMaster should just be executing commands at that layer.

With the existing abstraction, the flow is:
 - `addr` is "baked in" to an `I2CDevice` object at creation
 - Whenever the virtualizer wants an `I2CMaster` to do an operation for an `I2CDevice`, the virtualizer passes the relevant configuration for the current bus transaction to the `I2CMaster`

I think what gets weird, is that with this implementation the I2CMaster calls _up_ via the `set_address_check` to the virtualizer to ask the virtualizer if this is an okay address to set. This feels like it breaks an abstraction / layering we don't want to break.

Rather, I think that the `set_address_check` should occur whenever a request is made onto the `I2CMultiDevice`. _If_ the multidevice is attached to a virtualizer, it can ask the virtualizer itself if the requested address is legal. If there's no virtualizer (b/c MultiDevice is only I2C capsule on the board, say) then there's no one to ask, and that's something the MultiDevice should know and handle.

Then, the flow from the virtualizer's perspective can be the same. I'm doing an operation for this `I2C[Multi|Single]Device`, and I'll read from the chosen device the address to use and pass that down to the I2CMaster.

---

Hopefully that makes sense. It took me a while to wrap my head around what was bothering me with the layering, so it's entirely possible that what I wrote there isn't the clearest. I _think_ with that change, we'll have the final layering hurdle addressed and this can move forward.

**Comment by alistair23** [comment]
Hmm... I am misunderstanding the layering? I thought we have

```
--------------------------
| I2CMultiDevice |
--------------------------
|        MuxI2C       |
--------------------------
|     I2CMaster     |
--------------------------
```

Where the `I2CMultiDevice` calls the `set_address_check()` implemented by `MuxI2C`.

`MuxI2C` then has direct access to the physical `I2CMaster`.

That seems to follow the general abstraction layering, where a higher level capsule calls down and the upcalls go back up. Am I missing something?

**Comment by alistair23** [comment]
Any more comments?

**Comment by alistair23** [comment]
Ping!

**Comment by alistair23** [comment]
Ping!

**Comment by alistair23** [comment]
Ping!

**Comment by alistair23** [comment]
Ping!

**Comment by alistair23** [comment]
Ping!

**Comment by ppannuto** [comment]
Okay, re-reading this and paging things back in here; I think my [comment from the last review](https://github.com/tock/tock/pull/3867#issuecomment-2231900090) still stands, but I can maybe to try explain my thoughts a little better.

Copying the diagram from the reply to that comment, and annotating it a bit:

```--------------------------
| I2CMultiDevice |  I2CDevice  |  I2CDevice  | ...
        \              /           /
         \      /-----/           /
          \    /                 /
           \  / /---------------/
            \/-/
--------------------------
|     MuxI2C     |  <-- This is the layer responsible for managing the sharing of the physical I2C
                 | hardware. It knows all of the "kernel-owned" I2C addresses, by virtue of which
                 | I2CDevice(s) have registered with the mux. Critically, it's also responsible for
                 | managing what is currently active on the I2C bus.
                 | 
                 | I.e., if it's the MultiDevice's turn to access the physical hardware, then the Mux
                 | is responsible for configuring the physical bus with e.g. its new address
                 | _before_ allowing any commands through. By the time a `Cmd::Write`, passes
                 | through the Mux to the physical bus driver below, the write should be ready to
                 | go and require no more checks.
--------------------------
|    I2CMaster   |  <-- This is the layer responsible for controlling the physical bus. Only one thing
                 | can attach to a physical piece of hardware. Either some single thing that's the only
                 | bus user, or, more commonly some logic that allows for multiplexing the physical
                 | hardware. The important thing here is that this layer is *ignorant* of what's happening
                 | above it.
                 | 
                 | **I don't believe this PR should need to change anything in this layer.**
                 | 
                 | That's the 'cross-layering' issue I was getting at with where the 'set_address()' method
                 | ends up placed in the current code. I think it's the same thing Brad is trying to say with
                 | his comment here: https://github.com/tock/tock/pull/3867/files#r1548350963
--------------------------
```

**Comment by alistair23** [comment]
> **I don't believe this PR should need to change anything in this layer.**

Maybe I just haven't looked at this recently enough, but what changes to `I2CMaster` are there?

**Comment by alistair23** [comment]
The new HIL is

```rust
/// Extend the I2CDevice to add support for targetting multiple `I2CDevice`s.
pub trait I2CMultiDevice: I2CDevice {
    fn set_address(&self, addr: u8);
}
```

Which is implemented with

```rust
impl<'a, I: i2c::I2CMaster<'a>> i2c::I2CMultiDevice for I2CMultiDevice<'a, I, NoSMBus> {
    fn set_address(&self, addr: u8) {
        if addr == self.addr.get() {
            // Short circuit as address is already set
            return;
        }

        if self.mux.set_address_check(addr) {
            self.addr.set(addr);
        }
    }
}
```

So the implementation is in the device. The Mux layer provides a check, but `I2CMaster` is untouched. The `chips` directory (which implements `I2CMaster`) is unchanged except for the [OT divide by 0 bug fix](https://github.com/tock/tock/pull/3867/commits/fd348b5aa4f0f2f69b281b28ad749a88c01ddffc)
